### PR TITLE
Use already defined generateAssemblyReferenceSource rather defining a new target

### LINKED ISSUE #948: Unify functionality GenerateReferenceAssemblySource and GenerateReferenceSource
GenAPI defines GenerateReferenceAssemblySource https://github.com/dotnet/arcade/blob/55c0d49b9bec66981f7dd6d8f51a8dad43e8fbe0/src/Microsoft.DotNet.GenAPI/build/Microsoft.DotNet.GenAPI.targets#L31-L47, but we don't use it in CoreFx.

Instead we use GenerateReferenceSource https://github.com/dotnet/corefx/blob/4a84b8798dbd861675014b420e4c01f6a2116125/Directory.Build.targets#L176-L195.

We should delete GenerateReferenceSource and make sure that GenerateReferenceAssemblySource can handle our scenarios, then update the docs to to use it (eg: by setting the property).  

/cc @wtgodbe

### LINKED ISSUE #34604: Attempting to `GenerateReferenceSource` for System.Runtime fails
As per the title, the following fails:
```csharp
D:\tagoo\Repos\runtime\src\libraries\System.Runtime\src> dotnet msbuild /t:GenerateReferenceSource /p:Configuration=Release
Microsoft (R) Build Engine version 16.5.0+d4cbfca49 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Unable to resolve assembly 'Assembly(Name=System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)' referenced by 'Microsoft.Cci.DummyModule'
```

### PR DESCRIPTION
**Pull request by Anipik**
Fixes https://github.com/dotnet/runtime/issues/948
Fixes https://github.com/dotnet/runtime/issues/34604

### CODE REVIEW

**Comment by Anipik** [comment]
```
c:\git\runtime\src\libraries\System.Runtime\src>dotnet msbuild /t:GenerateReferenceAssemblySource /bl
Microsoft (R) Build Engine version 16.7.0-preview-20229-03+2cee6d020 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

C:\Program Files\dotnet\sdk\5.0.100-preview.5.20254.3\MSBuild.dll -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,C:\Program Files\dotnet\sdk\5.0.100-preview.5.20254.3\dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,C:\Program Files\dotnet\sdk\5.0.100-preview.5.20254.3\dotnet.dll -maxcpucount -verbosity:m /bl /t:GenerateReferenceAssemblySource .\System.Runtime.csproj
  Generated reference assembly source code: c:\git\runtime\src\libraries\System.Runtime\ref\System.Runtime.cs
```

File: src/libraries/Directory.Build.targets [hunk 1]
```diff
@@ -154,28 +154,24 @@
...
+    <_ExcludeAPIList>$([MSBuild]::NormalizePath('$(MSBuildProjectDirectory)', '..', 'ref', 'ReferenceSourceExcludeApi.txt'))</_ExcludeAPIList>
+    <_ExcludeAttributesList>$(RepositoryEngineeringDir)DefaultGenApiDocIds.txt</_ExcludeAttributesList>
+    <_LicenseHeaderTxtPath>$(RepositoryEngineeringDir)LicenseHeader.txt</_LicenseHeaderTxtPath> 
+    <GenAPILibPath>$(RefPath.Trim('\/'))</GenAPILibPath>
+    <GenAPITargetPath>$(_RefSourceFileOutputPath)</GenAPITargetPath>
+    <GenAPIAdditionalParameters>$(GenAPIAdditionalParameters) --exclude-attributes-list "$(_ExcludeAttributesList)"</GenAPIAdditionalParameters>
+    <GenAPIAdditionalParameters Condition="Exists('$(_ExcludeAPIList)')">$(GenAPIAdditionalParameters) --exclude-api-list "$(_ExcludeAPIList)"</GenAPIAdditionalParameters>
+    <GenAPIAdditionalParameters>$(GenAPIAdditionalParameters) --header-file "$(_LicenseHeaderTxtPath)"</GenAPIAdditionalParameters>
+    <GenAPIAdditionalParameters Condition="'$(LangVersion)' != ''">$(GenAPIAdditionalParameters) --lang-version "$(LangVersion)"</GenAPIAdditionalParameters> 
+    <ReferenceFromRuntimeCount>@(ReferenceFromRuntime->Count())</ReferenceFromRuntimeCount>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.targets:168]: Are you sure this works? Usually you can't access an item from a property unless you are inside a target.

File: src/libraries/Directory.Build.targets [hunk 2]
```diff
@@ -154,28 +154,24 @@
     <PackageReference Include="Microsoft.DotNet.GenFacades" Version="$(MicrosoftDotNetGenFacadesVersion)" PrivateAssets="all" IsImplicitlyDefined="true" />
   </ItemGroup>
 
-  <Target Name="GenerateReferenceSource">
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.targets:157]: This is a breaking change which should usually be part of a batched rollout. I'm unsure if we want to wait another month for this change to go in. cc @ericstj 

I assume docs need to be updated as well?

File: src/libraries/Directory.Build.targets (hunk 1 above)
Comment (Anipik) [reply src/libraries/Directory.Build.targets:168]: yeah my bad.

File: src/libraries/Directory.Build.targets [hunk 3]
```diff
@@ -154,28 +154,17 @@
...
-      <_GenAPICmd>$(_GenAPICmd) --exclude-attributes-list "$(_ExcludeAttributesList)"</_GenAPICmd>
-      <_GenAPICmd Condition="Exists('$(_ExcludeAPIList)')">$(_GenAPICmd) --exclude-api-list "$(_ExcludeAPIList)"</_GenAPICmd>
-      <_GenAPICmd>$(_GenAPICmd) --header-file "$(_LicenseHeaderTxtPath)"</_GenAPICmd>
-      <_GenAPICmd Condition="'$(LangVersion)' != ''">$(_GenAPICmd) --lang-version "$(LangVersion)"</_GenAPICmd>
-    </PropertyGroup>
-
-    <Exec Command="$(_GenAPICmd)" />
-    <Message Text="Generated reference assembly source code: $(_RefSourceFileOutputPath)" />
-  </Target>
+  <PropertyGroup>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.targets:157]: Should we set these properties unconditionally or as a BeforeTargets? Also GenAPI is only included for some assemblies: https://github.com/dotnet/runtime/blob/master/src/libraries/Directory.Build.targets#L153.

Comment (Anipik) [reply src/libraries/Directory.Build.targets:157]: If we really want to add a condition, i am more inclined towards adding the same condition to the property group rather than running a new target.

File: eng/referenceFromRuntime.targets [hunk 4]
```diff
@@ -111,6 +111,15 @@
     <Error Condition="'@(_missingReferenceFromRuntime)' != ''"
            Text="Could not resolve ReferenceFromRuntime item(s) '%(_missingReferenceFromRuntime.OriginalReferenceFromRuntime)' from '$(RuntimeProjectFile)'." />
   </Target>
+
+  <Target Name="SetGenApiProperties"
+          BeforeTargets="GenerateReferenceAssemblySource"
+          Condition="'@(ReferenceFromRuntime)' != ''">
+    <PropertyGroup>
+      <GenAPILibPath>$(RuntimePath.Trim('\/'));$(CoreCLRArtifactsPath.Trim('\/'))</GenAPILibPath>
```
Comment (ericstj) [inline eng/referenceFromRuntime.targets:119]: Why is this needed?  I would expect any references needed to already be passed via https://github.com/dotnet/arcade/blob/a8ee27e1afd19641e279207ac9117a4116afdd87/src/Microsoft.DotNet.GenAPI/build/Microsoft.DotNet.GenAPI.targets#L80-L82

Comment (Anipik) [reply eng/referenceFromRuntime.targets:119]: ReferencePath ItemGroup is not populated as GenerateReferenceAssemblySource does not depend upon resolve assembly reference. I have a pr up for that https://github.com/dotnet/arcade/pull/5419

**Comment by Anipik** [comment]
should wait until this **https://github.com/dotnet/arcade/pull/5419** is avaliable in runtime

**Comment by ViktorHofer** [comment]
I don't want to be over-aggressive with the batched rollout but you are renaming a target that devs are used to. That's a candidate for the next batched rollout: https://github.com/dotnet/runtime/issues/35895.

**Comment by ericstj** [comment]
You can keep the old target and just have it depend on the new one.  If you want to be cute you could emit a warning in the old target about deprecation.  Then delete the old one in next batched rollout.

**Comment by ViktorHofer** [comment]
I like cute.

**Comment by safern** [comment]
> You can keep the old target and just have it depend on the new one. If you want to be cute you could emit a warning in the old target about deprecation. Then delete the old one in next batched rollout.

I like this idea to not block until next batch rollout.

**Comment by Anipik** [comment]
i added the target back to unblock this and also updated the genapi version. I have tested the change locally and both the targets work

File: src/libraries/Directory.Build.targets [hunk 5]
```diff
@@ -154,27 +154,23 @@
...
+      <GenAPIAdditionalParameters>$(GenAPIAdditionalParameters) --header-file "$(_LicenseHeaderTxtPath)"</GenAPIAdditionalParameters>
+      <GenAPIAdditionalParameters Condition="'$(LangVersion)' != ''">$(GenAPIAdditionalParameters) --lang-version "$(LangVersion)"</GenAPIAdditionalParameters>  
+      <GenAPIAdditionalParameters Condition="'@(ReferenceFromRuntime)' != ''">$(GenAPIAdditionalParameters) --follow-type-forwards</GenAPIAdditionalParameters>
     </PropertyGroup>
+  </Target>
 
-    <Exec Command="$(_GenAPICmd)" />
-    <Message Text="Generated reference assembly source code: $(_RefSourceFileOutputPath)" />
+  <Target Name="GenerateReferenceSource" DependsOnTargets="GenerateReferenceAssemblySource">
+    <Message Text="This is a deprecated target. It will be removed in next infra rollout changes." Importance="high" />
```
Comment (safern) [inline src/libraries/Directory.Build.targets:173]: Add the new target name?

**Review by ViktorHofer — APPROVED** [review]

**Review by safern — APPROVED** [review]
